app/services/data_requests/lookup_shared_device_users.rb

```diff
@@ -9,66 +9,26 @@ module DataRequests
   # network are found.
   #
   class LookupSharedDeviceUsers
-    attr_reader :initial_users
+    attr_reader :initial_users, :depth
 
-    def initialize(initial_users)
+    def initialize(initial_users, depth = 3)
       @initial_users = initial_users
+      @depth = depth
+      @user_ids = initial_users.map(&:id)
+      @device_cookie_uuids = []
     end
 
     def call
-      lookup_users(initial_users)
-      loop do
-        break if user_uuids_to_lookup.empty?
-        lookup_users(users_to_lookup)
+      depth.times do |i|
+        warn "Searching at depth #{i}"
+        self.device_cookie_uuids = Device.where(user_id: user_ids).map(&:cookie_uuid).uniq
+        self.user_ids = Device.where(cookie_uuid: device_cookie_uuids).map(&:user_id).uniq
```

Comment:
1. Let's switch to `pluck()` so that we only pull the one string column out of the DB, and don't need to create entire AR models for each row in memoru
2. if these are sets we can use `+=` and it will do the uniqing for us

```suggestion
        self.device_cookie_uuids += Device.where(user_id: user_ids).pluck(:cookie_uuid)
        self.user_ids += Device.where(cookie_uuid: device_cookie_uuids).pluck(:user_id)
```